docs/architecture/adr-050-sign-mode-textual-annex1.md

```diff
@@ -0,0 +1,255 @@
+# ADR 050: SIGN_MODE_TEXTUAL: Annex 1 Value Renderers
+
+## Changelog
+
+- Dec 06, 2021: Initial Draft
+- Feb 07, 2022: Draft read and concept-ACKed by the Ledger team.
+
+## Status
+
+Draft. Concept ACK by the Ledger team, but details need to be polished.
+
+## Abstract
+
+This Annex describes value renderers, which are used for displaying Protobuf values in a human-friendly way using a string array.
+
+## Value Renderers
+
+Value Renderers describe how values of different Protobuf types should be encoded as a string array. Value renderers can be formalized as a set of bijective functions `func renderT(value T) []string`, where `T` is one of the below Protobuf types for which this spec is defined.
+
+### Protobuf `number`
+
+- Applies to:
+  - protobuf numeric integer types (`int{32,64}`, `uint{32,64}`, `sint{32,64}`, `fixed{32,64}`, `sfixed{32,64}`)
+  - strings whose `customtype` is `github.com/cosmos/cosmos-sdk/types.Int` or `github.com/cosmos/cosmos-sdk/types.Dec`
+  - bytes whose `customtype` is `github.com/cosmos/cosmos-sdk/types.Int` or `github.com/cosmos/cosmos-sdk/types.Dec`
+- Trailing zeroes are always removed
+- Formatting with `'`s for every three integral digits.
+- Usage of `.` to denote the decimal delimiter.
+
+#### Examples
+
+- `1000` (uint64) -> `["1'000"]`
+- `"1000000.00"` (string representing a Dec) -> `["1'000'000"]`
+- `"1000000.10"` (string representing a Dec) -> `["1'000'000.1"]`
+
+### `coin`
+
+- Applies to `cosmos.base.v1beta1.Coin`.
+- Denoms are converted to `display` denoms using `Metadata` (if available). **This requires a state query**. The definition of `Metadata` can be found in the [bank Protobuf definition](https://github.com/cosmos/cosmos-sdk/blob/v0.46.0-beta2/proto/cosmos/bank/v1beta1/bank.proto#L79-L108). If the `display` field is empty or nil, then we do not perform any denom conversion.
+- Amounts are converted to `display` denom amounts and rendered as `number`s above
+  - We do not change the capitalization of the denom. In practice, `display` denoms are stored in lowercase in state (e.g. `10 atom`), however they are often showed in UPPERCASE in everyday life (e.g. `10 ATOM`). Value renderers keep the case used in state, but we may recommend chains changing the denom metadata to be uppercase for better user display.
+- One space between the denom and amount
+- In the future, IBC denoms could maybe be converted to DID/IIDs, if we can find a robust way for doing this (ex. `cosmos:cosmos:hub:bank:denom:atom`)
+
+#### Examples
+
+- `1000000000uatom` -> `["1'000 atom"]`, because atom is the metadata's display denom.
+
+### `coins`
+
+TODO add spec
+
+### Example
+
+- `["2uatom", "3cosm"]` -> `"2 atom, 3 COSM"`
```

Comment:
Did you perhaps mean either `2000000uatom` on the left or `0.000'002 atom` on the right? (For consistency with line 47 above.) And note that if the left hand side is meant to denote the `Coins` as they'd be stored in a proto, they'd be alphabetical by denom, so `cosm` would come first. Decide whether the human-friendly representation should be in the order as stored, or whether it should be re-sorted to be lexicographic by the human-friendly name. The latter is less surprising to a human, but is one more processing step. Shrug.

Reply:
Thanks! I added some spec. Went with "it should be re-sorted to be lexicographic by the human-friendly name".